### PR TITLE
Base64 decoder rust/v11

### PR DESCRIPTION
**Pull request by inashivb**
Link to tickets:
- https://redmine.openinfosecfoundation.org/issues/7065
- https://redmine.openinfosecfoundation.org/issues/7058
- https://redmine.openinfosecfoundation.org/issues/6280

Previous PR: https://github.com/OISF/suricata/pull/11549

Changes since v10:
- combine rfc_2045 decoders, cleanups

### CODE REVIEW

File: rust/src/ffi/base64.rs [hunk 1]
```diff
@@ -26,6 +27,131 @@ pub enum Base64ReturnCode {
...
+            if !self.decoded.is_null() {
+                let _ = Vec::from_raw_parts(
+                    self.decoded,
+                    self.decoded_len as usize,
+                    self.decoded_len as usize,
+                );
+            }
+        }
+    }
+}
```
Comment (inashivb) [inline rust/src/ffi/base64.rs:101]: useless now. Should be cleaned up

Comment (inashivb) [reply rust/src/ffi/base64.rs:101]: This will increase code coverage a bit..

File: src/detect-transform-base64.c [hunk 2]
```diff
@@ -187,7 +182,7 @@ static int DetectTransformFromBase64DecodeTest01(void)
     PASS;
 }
 
-/* Simple success case with RFC2045 -- check buffer */
+/* Simple success case with Base64ModeBase64ModeRFC2045 -- check buffer */
```
Comment (inashivb) [inline src/detect-transform-base64.c:185]: typo

File: src/datasets.c [hunk 3]
```diff
@@ -518,13 +517,16 @@ static int DatasetLoadString(Dataset *set)
...
-            // coverity[alloc_strlen : FALSE]
             uint8_t decoded[strlen(line)];
-            uint32_t consumed = 0, num_decoded = 0;
-            Base64Ecode code = DecodeBase64(decoded, (uint32_t)strlen(line), (const uint8_t *)line,
-                    (uint32_t)strlen(line), &consumed, &num_decoded, Base64ModeStrict);
-            if (code == BASE64_ECODE_ERR) {
+            uint32_t num_decoded = 0;
+            bool b64d = Base64Decode((const uint8_t *)line, strlen(line), 0, Base64ModeStrict,
+                    decoded, &num_decoded);
+            if (b64d == false) {
```
Comment (jlucovsky) [inline src/datasets.c:524]: nit: suggest using `if (!b64d)` instead of comparing against `false`

File: src/datasets.c [hunk 4]
```diff
@@ -540,12 +542,16 @@ static int DatasetLoadString(Dataset *set)
...
```
Comment (jlucovsky) [inline src/datasets.c:549]: nit: ditto

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.

ERROR: QA failed on SURI_TLPR1_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 138 | 146 | 105.8% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 648 | 682 | 105.25% | 

Pipeline 21705

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.

ERROR: QA failed on SURI_TLPR1_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 138 | 146 | 105.8% |
| | SURI_TLPR1_stats_chk | | |
| .uptime | 648 | 671 | 103.55% | 

Pipeline 21707

File: src/detect-base64-decode.c [hunk 5]
```diff
@@ -87,19 +87,20 @@ int DetectBase64DecodeDoMatch(DetectEngineThreadCtx *det_ctx, const Signature *s
...
+        uint32_t num_decoded = 0;
+        (void)Base64Decode(payload, decode_len, det_ctx->base64_decoded_len_max, Base64ModeRFC4648,
+                det_ctx->base64_decoded, &num_decoded);
+        DEBUG_VALIDATE_BUG_ON(num_decoded >= decode_len);
+        det_ctx->base64_decoded_len = num_decoded;
+        SCLogDebug("Decoded %d bytes from base64 data.", det_ctx->base64_decoded_len);
+    }
 #if 0
     if (det_ctx->base64_decoded_len) {
         printf("Decoded data:\n");
```
Comment (jufajardini) [inline src/detect-base64-decode.c:106]: Can't we take advantage of this work to remove these `#if 0` portions, since there are clean-up commits?

**Review by jufajardini — COMMENTED** [review]
A few comments, but don't think any of them are blockers.

**Comment by ct0br0** [comment]
Just FYI - performance loss does seem consisent across runs. The uptime increase shows up in dev QA too.

File: rust/src/ffi/base64.rs [hunk 6]
```diff
@@ -26,6 +27,131 @@ pub enum Base64ReturnCode {
...
+/// It allows decoding in the modes described by ``Base64Mode`` enum.
+#[no_mangle]
+pub unsafe extern "C" fn Base64Decode(
+    input: *const u8, len: usize, max_decoded: u32, mode: Base64Mode, output: *mut u8, decoded_bytes: *mut u32
+) -> bool {
+    if input.is_null() || len == 0 {
+        return false;
+    }
+
+    let in_vec: &[u8] = std::slice::from_raw_parts(input as *mut u8, len);
```
[the remark this answers is not in the document]
Comment (inashivb) [reply rust/src/ffi/base64.rs:117]: indeed. thank you!

File: src/detect-base64-decode.c (hunk 5 above)
Comment (inashivb) [reply src/detect-base64-decode.c:106]: I'm often cautious about that as such debug stuff may come handy later on to debug stuff. Dead code, sure, should be removed. Commented debug statements, well, need more input on this..

**Comment by victorjulien** [comment]
> Just FYI - performance loss does seem consisent across runs. The uptime increase shows up in dev QA too.

Thanks Corey. This will need to be investigated.

Shivani, it will be a good idea to add 2 types of benchmarks for profiling:
1. very small data (to measure start up cost)
2. quite large data (to measure throughput)
Then inspect the performance and compare to the C implementation.

**Comment by inashivb** [comment]
> > Just FYI - performance loss does seem consisent across runs. The uptime increase shows up in dev QA too.
> 
> Thanks Corey. This will need to be investigated.
> 
> Shivani, it will be a good idea to add 2 types of benchmarks for profiling:
> 
>     1. very small data (to measure start up cost)
> 
>     2. quite large data (to measure throughput)
> 
> 
> Then inspect the performance and compare to the C implementation.

Thanks. Will do so. One thing I find weird is that this did not show up in https://github.com/OISF/suricata/pull/11544

File: src/detect-base64-decode.c (hunk 5 above)
Comment (jufajardini) [reply src/detect-base64-decode.c:106]: I see. Understood!

**Pull request closed without merging**